Abstracting NLP client

KamenDimitrov97 opened the pull request:
### What

I've added NLP client to dp-api-clients-go, point is to abstract the client so that it may be reused,
I've added new tests to test each call to a specific api
I've added mocks to mock the servers

### How to review

- [ ] Run tests - in nlp dir `goconvey` 
- [ ] Consistency
- [ ] Typos
- [ ] Comments
- [ ] Syntax order
- [x] Should the config live here?

### Who can review

Linden

lindenmckenzie commented on nlp/client.go:
```diff
@@ -0,0 +1,152 @@
+	client           dphttp.Client
+}
+
+// New creates a new elasticsearch client. Any trailing slashes from the URL are removed.
```
Don't think this is true.

lindenmckenzie commented on nlp/client.go:
```diff
@@ -0,0 +1,152 @@
+	"github.com/ONSdigital/log.go/v2/log"
+)
+
+// Client represents an instance of the elasticsearch client - now deprecated
```
Don't think this is true.

lindenmckenzie commented on nlp/client.go:
```diff
@@ -0,0 +1,152 @@
+		return category, err
+	}
+
+	log.Info(ctx, "successfully build category url", log.Data{"category url": categoryURL.String()})
```
This seems overly noisy to me.

lindenmckenzie reviewed (commented):
Some comments.

KamenDimitrov97 commented on nlp/client.go:
```diff
@@ -0,0 +1,152 @@
+package nlp
```
done

KamenDimitrov97 commented:
As per our review discussions 
I've fixed comments on the original PR to ONS repos
I've removed the common NLP client
I've created a separate Berlin client
I've created a separate Categories client
I've ran tests, created options and reusable interfaces

lindenmckenzie commented on nlp/berlin/README.md:
```diff
@@ -0,0 +1,63 @@
+
+// Add custom headers to the options
+options.Headers = http.Header{}
+options.Headers.Set(auithHeader, "")
```
authHeader?

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+		searchAPIClient := newBerlinAPIClient(t, httpClient)
+		check := initialTestState
+
+		Convey("When search API client Checker is called", func() {
```
not search API.

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+	Convey("Given clienter.Do returns an error", t, func() {
+		clientError := errors.New("unexpected error")
+		httpClient := newMockHTTPClient(&http.Response{}, clientError)
+		searchAPIClient := newBerlinAPIClient(t, httpClient)
```
not searchAPI

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+		searchAPIClient := newBerlinAPIClient(t, httpClient)
+		check := initialTestState
+
+		Convey("When search API client Checker is called", func() {
```
not search API

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+		check := initialTestState
+
+		Convey("When search API client Checker is called", func() {
+			err := searchAPIClient.Checker(ctx, &check)
```
not search API

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+
+	Convey("Given a 500 response for health check", t, func() {
+		httpClient := newMockHTTPClient(&http.Response{StatusCode: http.StatusInternalServerError}, nil)
+		searchAPIClient := newBerlinAPIClient(t, httpClient)
```
not searchAPI

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+			},
+			nil)
+
+		searchAPIClient := newBerlinAPIClient(t, httpClient)
```
not searchAPIClient

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+		Convey("When GetBerlin is called", func() {
+			query := url.Values{}
+			query.Add("q", "census")
+			resp, err := searchAPIClient.GetBerlin(ctx, Options{Query: query})
```
not searchAPIClient

lindenmckenzie commented on nlp/berlin/berlin_test.go:
```diff
@@ -0,0 +1,184 @@
+
+			Convey("Then the expected response body is returned", func() {
+				So(resp, ShouldNotBeNil)
+				So(resp.Matches[0].Codes[0], ShouldEqual, berlinResults.Matches[0].Codes[0])
+				So(resp.Matches[0].Encoding, ShouldEqual, berlinResults.Matches[0].Encoding)
+				So(resp.Matches[0].Names[0], ShouldEqual, berlinResults.Matches[0].Names[0])
+				So(resp.Matches[0].ID, ShouldEqual, berlinResults.Matches[0].ID)
+				So(resp.Matches[0].Key, ShouldEqual, berlinResults.Matches[0].Key)
+				So(resp.Matches[0].State[0], ShouldEqual, berlinResults.Matches[0].State[0])
+				So(resp.Matches[0].Subdivision[0], ShouldEqual, berlinResults.Matches[0].Subdivision[0])
+				So(resp.Matches[0].Words[0], ShouldEqual, berlinResults.Matches[0].Words[0])
```
could just use ShouldResemble here.

KamenDimitrov97 replied:
I know but shouldresemble didn't work for some reason and I got very frustrated and said screw it 
Now I retried it again and it worked I can't really say what I did differently

lindenmckenzie commented on nlp/berlin/options.go:
```diff
@@ -0,0 +1,32 @@
+	"net/url"
+)
+
+const (
+	// List of available headers
+	Authorization string = "Authorization"
+	CollectionID  string = "Collection-Id"
+)
```
what are you using these for? there is a headers/headers.go file which stores them.

lindenmckenzie commented on nlp/category/README.md:
```diff
@@ -0,0 +1,63 @@
+
+// Add custom headers to the options
+options.Headers = http.Header{}
+options.Headers.Set(auithHeader, "")
```
"authHeader"

lindenmckenzie commented on nlp/category/category.go:
```diff
@@ -0,0 +1,168 @@
+	healthcheck "github.com/ONSdigital/dp-api-clients-go/v2/health"
+	"github.com/ONSdigital/dp-api-clients-go/v2/nlp/category/models"
+	health "github.com/ONSdigital/dp-healthcheck/healthcheck"
+	"github.com/ONSdigital/dp-topic-api/sdk/errors"
```
why are you using dp-topic-api here?

lindenmckenzie reviewed (commented):
Few bits, happy to discuss.

KamenDimitrov97 commented on go.mod:
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/ONSdigital/dp-healthcheck v1.6.1
 	github.com/ONSdigital/dp-mocking v0.10.1
 	github.com/ONSdigital/dp-net/v2 v2.11.0
+	github.com/ONSdigital/dp-topic-api v0.21.0
```
Just noticed that and changed it, it was defaultly importing it because of the custom error

KamenDimitrov97 commented on nlp/berlin/berlin.go:
```diff
@@ -0,0 +1,168 @@
+
+	var req *http.Request
+
+	if payload != nil {
+		req, err = http.NewRequest(method, path, bytes.NewReader(payload))
+	} else {
+		req, err = http.NewRequest(method, path, http.NoBody)
+	}
```
I'll come back to this one

KamenDimitrov97 commented on nlp/README.md:
```diff
@@ -0,0 +1,27 @@
+TIncorporation of Natural Language Processing (NLP) into dp-search-api as a search engine. It's powered by three microservices: dp-berlin-api, dp-category-api, and dp-scrubber-api. 
+These services work together to upgrade the search experience. By utilizing NLP, the system delivers more accurate and intuitive search results, making searches more effective and user-friendly.
+
+Our aim is to optimize search results without compromising the current functionality.
```
It's on line 9
